### PR TITLE
Support negative numbers in readVLong

### PR DESCRIPTION
**Pull request by nik9000**
We don't *want* to use negative numbers with `writeVLong`
so throw an exception when we try. On the other
hand unforeseen bugs might cause us to write negative numbers (some versions of Elasticsearch don't have the exception, only an assertion)
so this fixes `readVLong` so that instead of reading a wrong
value and corrupting the stream it reads the negative value.

### CODE REVIEW

**Comment by nik9000** [comment]
Relates to #22285

**Comment by s1monw** [comment]
why don't we just make the assertion a hard check as well on the writer end?

**Comment by javanna** [comment]
I wonder if we should replace the assert with a proper exception instead? Would it have helped to catch this earlier if we had an exception instead?

**Comment by nik9000** [comment]
> why don't we just make the assertion a hard check as well on the writer end?

How about I throw an exception on the write but still preserve the ability to read negative vlongs so we don't get poisoned by any versions of Elasticsearch that don't have the fix?

**Comment by nik9000** [comment]
I pushed a change that makes the check on writes hard and drops the check on reads. That way if nodes with this code communicate with nodes without this code they'll be able to read the negative numbers.

File: core/src/test/java/org/elasticsearch/common/io/stream/BytesStreamsTests.java [hunk 1]
```diff
@@ -778,4 +777,27 @@ public void testReadNegativeArraySize() throws IOException {
...
+            output.writeVLongNoCheck(value);
+            StreamInput input = output.bytes().streamInput();
+            assertEquals(value, input.readVLong());
+        }
+        if (value < 0) {
+            BytesStreamOutput output = new BytesStreamOutput();
+            Exception e = expectThrows(IllegalStateException.class, () -> output.writeVLong(value));
+            assertEquals("Negative longs unsupported, use writeLong or writeZLong for negative numbers [" + value + "]", e.getMessage());
+        }
+    }
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/common/io/stream/BytesStreamsTests.java]: maybe we should also have one of those fun tests that read from a previous version to verify the interaction with previous versions?

Comment (nik9000) [inline core/src/test/java/org/elasticsearch/common/io/stream/BytesStreamsTests.java]: I think it is enough to use the `writeVLongNoCheck` version to put a negative number on the stream. At least, that was my intent.

Comment (javanna) [inline core/src/test/java/org/elasticsearch/common/io/stream/BytesStreamsTests.java]: ok that should repro the previous behaviour without assertions enabled you mean? I dunno, given that we are touching how we read and write stuff, I would be happy to have proper bw comp unit tests, especially around negative numbers that we don't cover anywhere in our test infra I guess.

**Comment by nik9000** [comment]
I pushed a commit that adds an assertion against some base 64 encoded bytes that'd be generated by 5.1.1 and below so we can be *sure* we can read from that version.

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java [hunk 2]
```diff
@@ -214,9 +214,8 @@ public long readLong() throws IOException {
     }
 
     /**
-     * Reads a long stored in variable-length format.  Reads between one and
-     * nine bytes.  Smaller values take fewer bytes.  Negative numbers are not
-     * supported.
+     * Reads a long stored in variable-length format. Reads between one and nine bytes. Smaller values take fewer bytes. Negative numbers
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: Comment: `one and nine` -> `one and ten`

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java [hunk 3]
```diff
@@ -260,8 +259,16 @@ public long readVLong() throws IOException {
...
-        return i | ((b & 0x7FL) << 56);
+        i |= ((b & 0x7FL) << 56);
+        if ((b & 0x80) == 0) {
+            return i;
+        }
+        b = readByte();
+        if (b != 0 && b != 1) {
+            throw new IOException("Invalid vlong (" + Integer.toHexString(b) + " << 63) | " + Long.toHexString(i));
+        }
+        i |= (b & 0x7FL) << 63;
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: I don't think the `&` is needed here, or is that just for symmetry?

**Review by jasontedor — COMMENTED** [review]
Looks good, I left two comments.

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java (hunk 2 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: Will fix.

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java (hunk 3 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: Mostly symmetry, but I can remove it, yeah.

**Comment by jasontedor** [comment]
I don't think there will be a 5.0.3 (and if there is, I don't think this should go into 5.0.3 otherwise it's too hard to reason about it being fixed in 5.0.3 but not 5.1.1).

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java (hunk 3 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: Oh, I'm fine with symmetry, I just wanted to make sure that I was reading it correctly.

**Review by jasontedor — APPROVED** [review]
LGTM.

**Comment by nik9000** [comment]
Thanks for reviewing @s1monw, @javanna, and @jasontedor!

master: 55099df1cb7f09e878a39f15592201c0f02cae4d
5.x:
5.1:

**Pull request merged**